### PR TITLE
Add allocation persistence

### PR DESCRIPTION
**Pull request by fabjan**
Problem
======

Projects can have "holes" with no variation or original allocation so that visitors end up outside the project. This is expected but this PHP SDK doesn't honor that.

The test configuration in the backend is mutable, but we want a visitor that have been allocated some variation (null or otherwise) to not change.

Solution
======

* Change the weighted allocation algorithm to be up to date with the Node.js and .NET SDKs and respect the project allocation "holes". The total weight is always 100 (and the backend ensures variations don't leak outside that in the config).

* Break out the cookie handling from `ensureVisitorID` to handle more concerns (allocations in addition to visitor ID), and move cookie writing to the end of the variation allocation.

* Use the JSON cookie to persist variation allocations made, and to reuse any persisted allocations before allocating anything.

* Add state information to projects and variations, so we can ignore inactive ones.

Testing
=====

This PR adds our data driven cross platform compatibility test suite, to ensure the observable behaviour is the same with regards to cookie input and output, and variation allocation based on hashing.

Verified that the examples still work.

### CODE REVIEW

File: src/Allocation.php [hunk 1]
```diff
@@ -17,67 +18,28 @@ final class Allocation
      *
      * If the visitor ID is empty, always returns the original variation.
      *
-     * @return VariationConfig the allocated variation
-     * @throws \Exception if the variation weight configuration in $project is invalid
+     * @return VariationConfig the allocated variation, or null
```
Comment (Hotas2k) [inline src/Allocation.php:21]: @return VariationConfig|null

File: src/Config/ProjectConfig.php [hunk 2]
```diff
@@ -30,8 +33,13 @@ function __construct(int $id, string $name, array $variations)
     public static function fromArray(array $data): ProjectConfig
     {
 
-        $id   = $data['id'] ?? 0;
-        $name = $data['name'] ?? '';
+        $id    = $data['id'] ?? 0;
+        $name  = $data['name'] ?? '';
+        $state = ProjectState::PAUSED;
+
+        if ('active' === ($data['state'] ?? null)) {
```
Comment (Hotas2k) [inline src/Config/ProjectConfig.php:40]: This is being done in multiple places, Perhaps we should move it into ProjectState? ProjectState::fromString(string $value)?

File: src/Config/VariationConfig.php [hunk 3]
```diff
@@ -30,8 +33,13 @@ public static function fromArray(array $data): VariationConfig
         $id     = $data['id'] ?? 0;
         $name   = $data['name'] ?? '';
         $weight = $data['weight'] ?? 1;
+        $state  = ProjectState::PAUSED;
```
Comment (Hotas2k) [inline src/Config/VariationConfig.php:36]: Is this ProjectState? Or do they just share the same options?

File: src/Allocation.php (hunk 1 above)
Comment (fabjan) [reply src/Allocation.php:21]: fixing

File: src/Config/ProjectConfig.php (hunk 2 above)
Comment (fabjan) [reply src/Config/ProjectConfig.php]: fixed

File: src/Config/VariationConfig.php (hunk 3 above)
Comment (fabjan) [reply src/Config/VariationConfig.php]: It's shared between them, I just didn't think of a good name. Let's go with RunState.

**Pull request merged**